### PR TITLE
feature(checksum): adding checksum validation

### PR DESCRIPTION
**Pull request by mboudreau**
Doing validation checksum of pact standalone binaries to make sure they are correct, allowing the use to specify a different URL or local path to the location of the binaries without compromising on the security.  This will fix this issue: https://github.com/pact-foundation/pact-node/issues/88

I still need to document the new environment variable available `PACT_BINARY_LOCATION`, which I might add a way to add it to package.json as a config as well.

### CODE REVIEW

File: standalone/install.ts [hunk 1]
```diff
@@ -60,31 +68,46 @@ function download(data: Data): Promise<Data> {
 					cd: `download-node-${data.platform}-${isCI ? "ci" : "user"}`
 				}
 			})
-				// Ignore all errors
-				.on("error", () => {});
+			// Ignore all errors
```
Comment (mefellows) [inline standalone/install.ts]: 👍

File: standalone/install.ts [hunk 2]
```diff
@@ -1,20 +1,28 @@
...
 const path = require("path");
 const fs = require("fs");
+const urljoin = require("url-join");
 const decompress = require("decompress");
 const tar = require("tar");
 const chalk = require("chalk");
+const sumchecker = require("sumchecker");
+
+export const PACT_STANDALONE_VERSION = "1.44.0";
+export const PACT_BINARY_LOCATION = process.env.PACT_BINARY_LOCATION ||
```
Comment (mefellows) [inline standalone/install.ts]: I thought we were going with the entry in `package.json` as per #88?

File: standalone/install.ts [hunk 3]
```diff
@@ -60,31 +68,46 @@ function download(data: Data): Promise<Data> {
...
+			// Or else it might be a local file, try to copy it over to the correct directory
+			fs.createReadStream(data.binaryDownloadPath)
+				.on("error", (e: string) => reject(`Error reading the file at '${data.binaryDownloadPath}': ${e}`))
+				.pipe(
+					fs.createWriteStream(data.filepath)
+						.on("error", (e: string) => reject(`Error writing the file to '${data.filepath}': ${e}`))
+						.on("close", () => resolve(data))
+				);
+		} else {
+			reject(`Could not get binary from locale ${data.binaryDownloadPath} as it's not a URL and does not exist at the path specified.`);
```
Comment (mefellows) [inline standalone/install.ts]: "Locale" or "local"?

File: standalone/install.ts (hunk 2 above)
Comment (mboudreau) [reply standalone/install.ts]: Yeah, still need to figure that out as I'm not sure how to get the package.json of the containing project...

File: standalone/install.ts (hunk 3 above)
Comment (mboudreau) [reply standalone/install.ts]: local :|

File: standalone/install.ts [hunk 4]
```diff
@@ -132,38 +167,84 @@ function extract(data: Data): Promise<void> {
...
+		platformFolderPath: path.resolve(__dirname, entry.folderName)
+	});
+}
+
+function join(...paths: string[]): string {
+	return HTTP_REGEX.test(paths[0]) ? urljoin(...paths) : path.join(...paths);
+}
+
+const checksumSuffix = ".checksum";
+const BINARIES: BinaryEntry[] = [
```
Comment (mefellows) [inline standalone/install.ts]: Nice, i think this makes the function a bit more readable.

File: standalone/install.ts (hunk 2 above)
Comment (mboudreau) [reply standalone/install.ts]: Not sure this is possible.  Seems that it only works if the thing runs through NPM itself, like (`npm run whatever`).  I might just stick to environment variables cause I think it just might be simpler for now.  We can always adapt it later.

**Comment by mboudreau** [comment]
na-na nah! super nerd!

**Comment by mboudreau** [comment]
Hm, apparently, only windows, pact-node isn't dispatching the 'started' event when starting the mock service.  Weird.  I wonder what changed.

**Comment by mboudreau** [comment]
Issue seems to be coming from the stub service.  I think there was a change recently that makes the stub service fail when setting a windows path to a local file.  Will need to fix the binary issue before this change can go out sadly.

**Pull request merged**